### PR TITLE
Add a pass to simplify single-character choices

### PR DESCRIPTION
**Pull request by markw65**
I have a modified version of examples/xml.peggy in my project, and was trying to optimize it for speed. I found a few places where it could be rewritten to avoid backtracking, getting about a 20% win, but with some readability cost.

Then I noticed that there are a number of long lists of alternate character classes, and they end up being implemented very inefficiently; effectively it tries one regex after another in sequence.

I manually combined all the character classes for BaseChar into a single character class, and got a nearly 2x speed up.

So I've added a pass that does that automatically. The result is a more than 3x speedup parsing the xml files that I work with (which seem fairly typical).

The pass takes a rule like `chars: "a" / "b" / [d-r]` and turns it into `chars: [abd-r]`

### CODE REVIEW

**Comment by markw65** [comment]
I've put it under an option, `mergeCharacterClasses` because enabling it by default breaks a number of tests, which have specific expectations for error messages:

```
    Expected value to strictly be equal to:
      "Expected \"a\", \"b\", or \"c\" but \"d\" found."
    Received:
      "Expected [abc] but \"d\" found."
```

Since that test is specifically designed to ensure that an appropriate error message is output for `choice` nodes I didn't want to just change the expected output. But I'd be happy to change the grammar for such tests so that the choice nodes remain as choice nodes, and enable the optimization by default if that would be preferable.

**Review by Mingun — CHANGES_REQUESTED** [review]
I generally agree, that such optimization is very useful, but we need more tests to ensure, that it works as expected and does not processes pieces which it shouldn't.

Can you please add tests for negative scenarios, at least.

I'm not sure, should we add a setting for this optimization and should it be enabled by default. It seems to me, that settings is superfluous and we can apply this optimization always. So yes, I'm fine with changing expected error messages in failed tests

File: test/unit/compiler/passes/merge-character-classes.spec.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+const expect = chai.expect;
+
+describe("compiler pass |mergeCharacterClasses|", () => {
+  it("Merges single character literals, class, and ref_rules", () => {
+    expect(pass).to.changeAST(
+      [
+        "one = three / 'a' / [c-g]",
+        "two = three / 'P' / [Q-T]",
+        "three = 'x' / [u-w]",
```
[the remark this answers is not in the document]
Comment (markw65) [reply test/unit/compiler/passes/merge-character-classes.spec.js:17]: Mostly agreed - I'll certainly add more tests.

> For example, if three rule will contain $ operator, or action block

In that case, its `expression`'s  type will be `text` or `action` and `asClass` will ignore it. So I don't need to do anything to handle that.

But you made me realize that I should explicitly check for `text` and look through it, because we actually want 
`foo = "a" / $("b" / [xyz]) / "X"` to get converted to `foo = [Xabxyz]`.

**Comment by markw65** [comment]
I just realized I need to update `lib/peggy.js` and `docs/js/examples.js` now that this is done unconditionally. Should I do that now, or wait until you're happy with the state of the commit?

**Comment by hildjj** [comment]
This looks good to me, I think.  @Mingun any other comments?

**Comment by hildjj** [comment]
Except one needed addition:  a CHANGELOG entry, please.

**Comment by markw65** [comment]
> Except one needed addition: a CHANGELOG entry, please

I put it in "minor" updates - let me know if if should be somewhere else...

**Comment by hildjj** [comment]
Sorry for moving the goalposts, but once I thought of coverage on #427, there are a few lines that need to be hit on this one in lib/compiler/passes/merge-character-classes.js, unless they are particularly hard to reach.

File: lib/compiler/passes/merge-character-classes.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  const [asClass, merge] = [
+    node => {
+      if (node.type === "class" && !node.inverted) {
+        return node;
+      }
+      if (node.type === "text") {
+        return asClass(node.expression);
+      }
+      if (node.type === "literal" && node.value.length === 1) {
+        return { type: "class", parts: [node.value], inverted:false, ignoreCase: node.ignoreCase };
```
Comment (Mingun) [inline lib/compiler/passes/merge-character-classes.js:70]:
```suggestion
        return { type: "class", parts: [node.value], inverted: false, ignoreCase: node.ignoreCase };
```

**Review by Mingun — APPROVED** [review]
Need to check how this optimization will affect source map generation. Other than that, all is good

File: lib/compiler/passes/merge-character-classes.js (hunk 2 above)
Comment (markw65) [reply lib/compiler/passes/merge-character-classes.js:70]: > Need to check how this optimization will affect source map generation

Good point. I think there's one small issue I need to address...

**Comment by hildjj** [comment]
I'm going off what I see in the `coverage/lcov-report` directory after having run `npm run build`.  That dir can sometimes be a little confusing because it doesn't get cleaned up every time, and the URLs change if you rerun with a subset of the tests.  Try deleting the directory, then navigating to the correct file from `index.html` when you run a different set of tests.

Here are the three lines that I currently see uncovered in the full test suite: 22, 67, 73.

I don't see breakpoints hit on any of those, but checked to make sure that I have everything set up right by having breakpoints fire on other lines in your file.

**Comment by hildjj** [comment]
Also, I'm on the Discord channel if you want to try and figure this out in realtime.

**Comment by markw65** [comment]
I rewrote things a little:
 - some things that asClass was checking for were unnecessary because merge would have already handled them
 - preserve locations when merging nodes
 - only call cleanParts after merging all the alternates for efficiency

And then I updated the tests so that everything is covered.

**Comment by hildjj** [comment]
@Mingun, one last thumbs-up, please?

**Comment by hildjj** [comment]
@Mingun this one needs a final thumbs-up as well, please.

File: lib/compiler/passes/merge-character-classes.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+          const cls = asClass(ref);
+          // Return a clone, not the original, because the value we return
+          // may get merged into.
+          return cls && {
+            type: cls.type,
+            parts: cls.parts.slice(),
+            inverted: cls.inverted,
+            ignoreCase: cls.ignoreCase,
+            location: node.location,
+          };
```
Comment (Mingun) [inline lib/compiler/passes/merge-character-classes.js:91]: Maybe it could be better to request cloned data from `asClass`, because we need to clone only `class` nodes, other nodes already created by us as new objects.

So probably
```javascript
return asClass(ref, true/* return a clone */);
```
here and `asClass(..., false)` otherwise else. Or even make clone at line 118
```javascript
prev = cls;
```
It seems the clone is really needed only there

File: test/behavior/generated-parser-behavior.spec.js [hunk 4]
```diff
@@ -2677,13 +2677,13 @@ describe("generated parser behavior", () => {
...
 
           expect(parser).to.failToParse("d", {
             expected: [
-              { type: "literal", text: "a", ignoreCase: false },
-              { type: "literal", text: "b", ignoreCase: false },
-              { type: "literal", text: "c", ignoreCase: false },
+              { type: "literal", text: "aa", ignoreCase: false },
+              { type: "literal", text: "bb", ignoreCase: false },
+              { type: "literal", text: "cc", ignoreCase: false },
             ],
```
Comment (Mingun) [inline test/behavior/generated-parser-behavior.spec.js:2687]: Strictly speaking, we should give parser a chance to not fail by parsing `dd` here and below :)

File: test/unit/compiler/passes/merge-character-classes.spec.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            name: "three",
+            expression: {
+              type: "class",
+              parts: [["u", "x"]],
+              ignoreCase: false,
+              inverted: false,
+              location: {
+                start: { line: 3, column: 11 },
+                end: { line: 3, column: 22 },
+              },
```
Comment (Mingun) [inline test/unit/compiler/passes/merge-character-classes.spec.js:58]: This location covers
```pegjs
'x' / [u-w]
```
Shouldn't it cover
```pegjs
$('x' / [u-w])
```
?

File: test/unit/compiler/passes/merge-character-classes.spec.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+const expect = chai.expect;
+
+describe("compiler pass |mergeCharacterClasses|", () => {
+  it("Merges single character literals, class, and ref_rules", () => {
+    expect(pass).to.changeAST(
+      [
+        "one = three / $('a' / 'd') / [c-f] / [efh] / [c-h] / [c-g]",
+        "two = 'P' / 'P' / three / 'P' / [Q-T]",
+        "three = $('x' / [u-w])",
+        "four = 'a' / [aaa] / 'a'",
```
Comment (Mingun) [inline test/unit/compiler/passes/merge-character-classes.spec.js:18]: Because merging also should work when all literal / classes ignores case, I suggest copy the whole test and add `i` suffix here to test `ignoreCase` case

**Review by Mingun — APPROVED** [review]
My approval, except minor nits.

It seems we can slightly optimize by reducing number of unnecessary clones, but benefit also seems to be very low. So can leave this in current state.

We can add tests for `ignoreCase` cases.

Also I suggest to insert `session.emitInfo()` calls when node changed.

File: lib/compiler/passes/merge-character-classes.js (hunk 3 above)
Comment (markw65) [reply lib/compiler/passes/merge-character-classes.js:91]: Makes sense.

While we're talking about clones though, it seems to be hard to produce generic clones in this project.

If I say `Object.assign({}, objectToClone)` eslint tells me to use `{ ...objectToClone }`, but if I say `{ ...objectToClone }`, it tells me `Parsing error: Unexpected token ...`.

I don't see why we shouldn't use `...` (it works with node14 and later, and will presumably get fixed by rollup)? Can I change the eslint settings?

File: test/unit/compiler/passes/merge-character-classes.spec.js (hunk 5 above)
Comment (markw65) [reply test/unit/compiler/passes/merge-character-classes.spec.js:58]: will fix

File: lib/compiler/passes/merge-character-classes.js (hunk 3 above)
Comment (hildjj) [reply lib/compiler/passes/merge-character-classes.js:91]: If you double-check that `...` gets replaced by the rollup step, submit a bug to https://github.com/peggyjs/peggyjs-eslint-config and I'll fix it.  There are a couple of other things I want to make sure are up-to-date while I'm in there.

Comment (hildjj) [reply lib/compiler/passes/merge-character-classes.js:91]: (it's actually the typescript step that is likely doing the replacement, which will be easier to check)

**Comment by markw65** [comment]
Anything else to do here?

**Comment by markw65** [comment]
@hildjj @Mingun ping again?

**Review by Mingun — APPROVED** [review]
